crates/sui-framework/packages/sui-framework/sources/random.move

```diff
@@ -122,4 +126,124 @@ module sui::random {
+        g.counter = g.counter + 1;
+        // TODO[crypto]: should blake be used here instead?
+        hmac_sha3_256(&g.seed, &bcs::to_bytes(&g.counter))
+    }
+
+    // Fill the generator's buffer with 32 random bytes.
+    fun fill_buffer(g: &mut RandomGenerator) {
+        let next_block = derive_next_block(g);
+        vector::append(&mut g.buffer, next_block);
+    }
+
+    /// Generate n random bytes.
+    public fun bytes(g: &mut RandomGenerator, num_of_bytes: u16): vector<u8> {
+        // TODO[move]: should we have this limit or just leave it to gas limit?
+        assert!(num_of_bytes <= MAX_RANDOM_BYTES, ETooManyBytes);
+        let result = vector::empty();
+        // Append RAND_OUTPUT_LEN size buffers directly without going through the generator's buffer.
+        let num_of_blocks = num_of_bytes / RAND_OUTPUT_LEN;
+        while (num_of_blocks > 0) {
+            vector::append(&mut result, derive_next_block(g));
+            num_of_blocks = num_of_blocks - 1;
+        };
+        // Take remaining bytes from the generator's buffer.
+        if (vector::length(&g.buffer) < ((num_of_bytes as u64) - vector::length(&result))) {
+            fill_buffer(g);
+        };
+        while (vector::length(&result) < (num_of_bytes as u64)) {
+            vector::push_back(&mut result, vector::pop_back(&mut g.buffer));
+        };
+        result
+    }
+
+    // Helper function that extracts the given number of bytes from the random generator and returns it as u256.
+    // Assumes that the caller has already checked that num_of_bytes is valid.
+    fun u256_from_bytes(g: &mut RandomGenerator, num_of_bytes: u8): u256 {
+        if (vector::length(&g.buffer) < (num_of_bytes as u64)) {
+            fill_buffer(g);
+        };
+        let result: u256 = 0;
+        let i = 0;
+        while (i < num_of_bytes) {
+            let byte = vector::pop_back(&mut g.buffer);
+            result = (result << 8) + (byte as u256);
+            i = i + 1;
+        };
+        result
+    }
+
+    /// Generate a u256.
+    public fun generate_u256(g: &mut RandomGenerator): u256 {
+        u256_from_bytes(g, 32)
+    }
+
+    /// Generate a u128.
+    public fun generate_u128(g: &mut RandomGenerator): u128 {
+        (u256_from_bytes(g, 16) as u128)
+    }
+
+    /// Generate a u64.
+    public fun generate_u64(g: &mut RandomGenerator): u64 {
+        (u256_from_bytes(g, 8) as u64)
+    }
+
+    /// Generate a u32.
+    public fun generate_u32(g: &mut RandomGenerator): u32 {
+        (u256_from_bytes(g, 4) as u32)
+    }
+
+    /// Generate a u16.
+    public fun generate_u16(g: &mut RandomGenerator): u16 {
+        (u256_from_bytes(g, 2) as u16)
+    }
+
+    /// Generate a u8.
+    public fun generate_u8(g: &mut RandomGenerator): u8 {
+        (u256_from_bytes(g, 1) as u8)
+    }
+
+    /// Generate a random u128 in [min, max] (with a bias of 2^{-128}).
+    public fun generate_u128_in_range(g: &mut RandomGenerator, min: u128, max: u128): u128 {
```

Comment:
Why just for u128?

Reply:
Added the rest except for u256 which we cannot implement purely in Move (since we need larger than u256 integers). We could add a native function in the future for that if needed.